### ISSUE TITLE
Core doesn't boot

### ISSUE THREAD

**Issue by Akaifutari** [MiSTer-devel/Arcade-Cave_MiSTer#18]
Hi Josh!

I have an issue with the new beta version, the game doesn't boot. I have a black screen all the time.
My Mister have 128Mb module. It could be the problem?

Thanks!

**Comment by nullobject** [comment]
@Akaifutari I'm sorry it's not working. You are right, it is likely to be a SDRAM timing problem.

Do you have any other SDRAM modules you could test it with (e.g. 32MB)?

**Comment by Akaifutari** [comment]
Nope, I don't have any other SDRAM module.

No problem, I'll wait for the fix.

**Comment by sh00tthec0re** [comment]
I had no problem running the previous betas but Beta 6 hard locks my MiSTer immediately at core launch - the "Kbd Mode" message in the upper-left corner stays visible and the controls are completely unresponsive.  I have the hit the first button on the MiSTer to restart the machine.

I'm using a 128 meg SDRAM module that tests stable at 140mhz (but no faster).  I'm using VGA output.  I tried different vsync_mode options in mister.ini but the outcome is the same.
![85FYnMa8](https://user-images.githubusercontent.com/49319761/104115444-c4977b80-52cc-11eb-9f1f-f4e7c74e04b6.jpeg)

**Comment by Akaifutari** [comment]
I have news about this. I tried beta 6 for Dodonpachi, Dodonpachi (Japan) and Dodonpachi (Arrange) and still doesn't boot, but Dodonpachi (No Warn) works like a charm.

I tried the new Arcade-Cave_DDR rbf file and works flawless.

**Comment by nullobject** [comment]
@Akaifutari Really good to know, thanks for reporting in :+1:

**Comment by nullobject** [comment]
@sh00tthec0re When you get a chance, can you please try the DDR-only version and let me know what happens.

**Comment by sh00tthec0re** [comment]
I deleted all prior DDP Beta RBF and MRA files, then installed the DDP DDR-only RBF core file and the Beta 6 MRA files.  Unfortunately I'm seeing the exact same freezing up on core load behavior as I saw with the Beta 6 RBF core.  I tried all four MRA files and there's no difference - including No Warn.

**Comment by Akaifutari** [comment]
@sh00tthec0re Have you wait a little when you load the core? I noticed that this core need to wait a few more seconds than the other cores till the first screen appear.

**Comment by sh00tthec0re** [comment]
@Akaifutari yes, I've let it sit for a full minute with no change.  The previous Betas would boot up to the legal warning after about 20 seconds and you'd see some movement in the graphical "garbage" as it booted.  Beta 6 doesn't show any garbage movement at all and the MiSTer completely locks up.

**Comment by nullobject** [comment]
@sh00tthec0re That's a shame. I will keep looking for what could be causing this issue :eyes:

**Comment by calle81** [comment]
I have the same exact issue as sh00tthec0re. My Mister completely hangs and I cannot bring up the OSD as soon as I launch any of the MRAs. No problem with CPS 1.5 or Jotego's RAM tests.

**Comment by ercan1441** [comment]
> I have the same exact issue as sh00tthec0re. My Mister completely hangs and I cannot bring up the OSD as soon as I launch any of the MRAs. No problem with CPS 1.5 or Jotego's RAM tests.

Same issue here, complete lockup. Also using a 128MB SDRAM which tested succesfully with Jotego’s latest SDRAM tester.

**Comment by geekmiki** [comment]
Same here as sh00tthec0re, calle81 and ercan1441. I managed to boot it once though, randomly after dozens of failures. Hasn't worked since.
Tried 60hz mode with no luck.

**Comment by coderkind** [comment]
Same here; 128MB RAM module, and tried all the available versions of Dodonpachi available within the MiSTer menu. All cause the MiSTer to freeze on a similar screen to the one above ("Kbd mode: Joystick 1"). MiSTer has previously worked fine with Jotego's CPS 1.5 core.

**Comment by sh00tthec0re** [comment]
Public core started working for me after I deleted and recopied the DDP MAME ZIP file to /games/mame/

**Comment by nullobject** [comment]
> Public core started working for me after I deleted and recopied the DDP MAME ZIP file to /games/mame/

@sh00tthec0re Winner!

**Comment by calle81** [comment]
> > Public core started working for me after I deleted and recopied the DDP MAME ZIP file to /games/mame/
> 
> @sh00tthec0re Winner!

I have the rom with that checksum but still the same problem. 
![image](https://user-images.githubusercontent.com/7829254/104526150-080e2480-5602-11eb-9d0b-82b1a8b6f089.png)

**Comment by ItsAtomico** [comment]
Issue still persists for me. Black screen with "kbd mode" sign, DE10 locks up.
MD5 checks out (.227 rom). Also tried the .221 set and whatever version the update script pulled.  joystick 1". 

As for the rbf and mras, used both the official and unofficial updaters, as well as manually installing the files from the patreon post. 

Connected my mister via serial earlier, got this log, in case it helps. 
[ddp_error.txt](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/files/5811667/ddp_error.txt)

128MB, I/O 6.0, OTG hub.

Edit: I loaded the core after deleting the rom to get to the menu, switched to DDR3, still no dice.

**Comment by johnklax** [comment]
I'm having the exact same problem as ItsAtomico and followed the same trouble shooting steps except for deleting the rom to get into the menu. I verified that the rom file matches the checksum above.

I'm using: 128 MB, Digital I/O 1.2, USB Hub 2.1

**Comment by stevep555** [comment]
The locking up issue - I think it's related to the resolution (video_mode) setting in combination with having vscale_mode set to integer scaling.

At 1280 x 720 @ 60Hz and integer scaling set (vscale_mode=1), I get lockups.  But if I try a different resolution or set vscale_mode to 0 (for instance), it works fine.

**Comment by Akaifutari** [comment]
The public release run perfectly now for me. I have no problems with Dodonpachi, Dodonpachi (Japan) and Dodonpachi (No Warn)

**Comment by kiFlaFlaki** [comment]
Same problem, MiSTer locks on kbd screen. Also 128Mb. Switched to 60Hz, DDR, same. All DDP versions locks.

**Comment by ItsAtomico** [comment]
Switched resolutions and scaling, no luck.

**Comment by kiFlaFlaki** [comment]
Also, switched resolutions and scaling, same problem.

**Comment by alienator303** [comment]
Same problem here, after running the update_all script, latest rbf crashes the Mister , all beta's before worked perfect. help

**Comment by jlrh** [comment]
Same problem here. Black screen with joystick msg and core menu button not responding.

I use 32MB SDRAM. 

I've tried to change SDRAM to DDR3, 57,5Hz to 60Hz, disable all sprites and layers in debug menu, different ROMS... without results.

The japan version worked only one time. I exited, and it didn't work anymore.

I've tried two different SD cards too (128GB and 32GB sandisk ultra).

**Comment by ratherDashing** [comment]
+1.  Never had any problems with any other cores.  128 SDRAM from MiSTERaddons.  Used the update_all script.  Tried all MRAs with no mister.ini file (so all defaults).  All of them freeze the device and need to be unplugged.

**Comment by bellwood420** [comment]
I would like to share my result.

I tried many times, and finally I found that the issue occurs rarely and randomly in my environment.
And it doesn't seem to be related to settings like resolution or vscale_mode.
It's mostly fine, but sometimes it doesn't boot just like all of you.

**Environment and settings.**
- 128 MB SDRAM
- 1280x720@60/1920x1080@60
- HDMI output to 2K Display
- vscale_mode=0/1
- ddonpach.zip (58a82638a2aa7e4e40b32bf365bff17d)
- ddonpachj.zip (a763d6bb37cc6aad573f6da419157a07)
- MiSTer v210109, OS v200908

Serial log shows:
```
arcade_root /media/fat/_Arcade
mame_root /media/fat/_Arcade
Using interleave: input 8, output 16
        file: /media/fat/_Arcade/mame/ddonpach.zip/u27.bin, start=0, len=0, map(2)=10
     0: file: /media/fat/_Arcade/mame/ddonpach.zip/u26.bin, start=0, len=0, map(2)=1
Disable interleave
100000: file: /media/fat/_Arcade/mame/ddonpach.zip/u50.bin, start=0, len=0
300000: file: /media/fat/_Arcade/mame/ddonpach.zip/u51.bin, start=0, len=0
500000: file: /media/fat/_Arcade/mame/ddonpach.zip/u52.bin, start=0, len=0
700000: file: /media/fat/_Arcade/mame/ddonpach.zip/u53.bin, start=0, len=0
900000: file: /media/fat/_Arcade/mame/ddonpach.zip/u60.bin, start=0, len=0
B00000: file: /media/fat/_Arcade/mame/ddonpach.zip/u61.bin, start=0, len=0
D00000: file: /media/fat/_Arcade/mame/ddonpach.zip/u62.bin, start=0, len=0
F00000: file: /media/fat/_Arcade/mame/ddonpach.zip/u6.bin, start=0, len=0
1100000: file: /media/fat/_Arcade/mame/ddonpach.zip/u7.bin, start=0, len=0
file_finish: 0x1300000 bytes sent to FPGA
```

When it freezes, the last line (file_finish) does not appear.
The last line is missing in @ItsAtomico  's log, too.

Something wrong is happening there.

**Comment by geekmiki** [comment]
Same here, worked consistently yesterday after changing the rom to ddonpach.zip (58a82638a2aa7e4e40b32bf365bff17d), but doesn't work today on the exact same setup.

**Comment by nullobject** [comment]
> Same problem here, after running the update_all script, latest rbf crashes the Mister , all beta's before worked perfect. help

@alienator303 It's interesting that all the betas worked :thinking:

> When it freezes, the last line (file_finish) does not appear.

@bellwood420 Thank you for the details, it's quite interesting. The hard MiSTer lockup during boot is starting to make me think there is something wrong with the boot sequence.

I'm am investigating this today...

**Comment by nullobject** [comment]
> Same here, worked consistently yesterday after changing the rom to ddonpach.zip (58a82638a2aa7e4e40b32bf365bff17d), but doesn't work today on the exact same setup.

I wish it was unreliable for me too, then I could quite easily find what the problem was :smile:

**Comment by nullobject** [comment]
Including @mfncl9991's comment from a duplicate issue:

> Some of us over on classic gaming discord are seeing issues where DoDonPachi public release freezes with a black screen showing "Kbd mode: Joystick 1".
> 
> I ran update_all to pull the files initially, and confirmed the MRAs and core RBF are where they should be (/media/fat/_Arcade/cores) , and I see the ROM zip in /media/fat/games/mame.
> 
> /media/fat# ls -la "_Arcade" | grep DoDon
> -rwxr-xr-x 1 root root 1121 Jan 13 22:07 DoDonPachi (Arrange).mra
> -rwxr-xr-x 1 root root 1157 Jan 13 22:07 DoDonPachi (Japan).mra
> -rwxr-xr-x 1 root root 1133 Jan 13 22:07 DoDonPachi.mra
> -rwxr-xr-x 1 root root 1362 Jan 13 22:07 DoDonPachi (no warn).mra
> 
> /media/fat# ls -la "_Arcade/cores" | grep Cave
> -rwxr-xr-x 1 root root 3888860 Jan 14 10:57 Arcade-Cave_20210113.rbf
> -rwxr-xr-x 1 root root 3888860 Jan 13 22:07 Cave_20210113.rbf
> 
> /media/fat# ls -la "games/mame" | grep ddon
> -rwxr-xr-x 1 root root 9734822 Sep 8 12:40 ddonpach.zip
> 
> /media/fat# md5sum games/mame/ddonpach.zip
> 58a82638a2aa7e4e40b32bf365bff17d games/mame/ddonpach.zip
> 
> Tried vsync_mode=0 but same behavior, suggesting this is not a display issue.
> As a Patron I had beta 2 working in the past. I also ran jotego's 128MB module memory tester, and all passed OK, so I don't think its an SDRAM issue (CPS1.5 core working fine).
> 
> Update: also tried the DDR mode RBF, and the same freeze happens. This time instead of a black screen, I have a garbled frozen display (also with "Kbd mode: Joystick 1" status message frozen)

**Comment by nullobject** [comment]
Ok, my `ddonpach.zip` is different to the one you guys seem to be using:

```
/media/fat/games/mame# md5sum ddonpach.zip 
9f887a16b6645327384a6623cbebb807  ddonpach.zip
```

Can somebody please hook me up with the 58a82638a2aa7e4e40b32bf365bff17d one? I need to try to get as close to your setup as possible.

**Comment by johnklax** [comment]
Thanks Josh!
Just sent you an email!

**Comment by nullobject** [comment]
Next question: does the yellow LED go off after the core freezes, or does it just stay on?

![image](https://user-images.githubusercontent.com/3614/104666367-cecfd480-5727-11eb-87c7-49665e7a90ac.jpeg)

**Comment by nullobject** [comment]
@johnklax Interesting. The user (third/yellow) LED should turn _off_ when the core has finished loading.

This is good to know :thinking:

**Comment by bellwood420** [comment]
The issue also happened with 9f887a16b6645327384a6623cbebb807 ddonpach.zip

As you may have already checked,  base binaries inside are the same.

**Comment by mfncl9991** [comment]
When I moved the mister to my CRT and tried it there, the core loaded once and I was able to play no issues. The next time (and all the times afterwards) it froze at the same spot during loading as it did when connected to an HDMI display.

**Comment by nullobject** [comment]
Just for my own sanity, when you guys get a chance could you please try the DDR-only version of the core (this was from beta 6).

This is just to 100% rule out any SDRAM issues.  Let me know if you tried it and it didn't work.

(You will have to move the other Cave RBF files out of the way).

https://1drv.ms/u/s!AlvbZtqpJbLwgiTApdbfXHWXKJ0f?e=QHJMBp

**Comment by MJY71** [comment]
What I thought was odd was that the patched (No Warn) MRA was loading just fine when the unpatched International, Master Ver. 970205 MRA would not. Other than those patches, they are practically identical.

I took the `<skip CRC>` patch (`<patch offset="0x005404">00 60 0C 00</patch>`) from the No Warn MRA and carried it over to the International MRA, and it now that one boots just fine. For the Japan, Master Ver. 970205 MRA, that patch has no effect... I'm guessing because the offset needs to be changed since the main CPU ROMs are different.

The Arrange Ver MRA seems fine and requires no additional modification.

**Comment by nullobject** [comment]
Next experiment: this core will enable the green (disk) LED while the DDR3 memory is waiting for data to be written. It should turn _off_ when the core has finished loading.

https://1drv.ms/u/s!AlvbZtqpJbLwgiWBWCI0rGuFRFGk?e=EgNgsD

**Comment by bellwood420** [comment]
I tried Arcade-Cave_DDR_20210110.rbf but the issue still exists. Rather, probability of freeze increased.

I don't have I/O board but, the same serial log shows that the HPS does not reach the following line.
Maybe stucking somewhere.

https://github.com/MiSTer-devel/Main_MiSTer/blob/4092fb61ae8b314c51b460a18e28ed5d2440286f/support/arcade/mra_loader.cpp#L353

**Comment by georam5** [comment]
> Next experiment: this core will enable the green (disk) LED while the DDR3/SDRAM memory is waiting for data to be written. It should turn _off_ once the core has finished loading.
> 
> https://1drv.ms/u/s!AlvbZtqpJbLwgiWBWCI0rGuFRFGk?e=EgNgsD

This worked for me! I have the 128 Meg SD ram as well and was getting the freeze immediately with kbd mode stuck in upper left since yesterday after multiple tries. As soon as I added this rbf it worked great played through a few stages and everything seems to be working as it should. Thank you for working on this so quickly, really enjoying the core!

**Comment by bellwood420** [comment]
Sorry, I missed Arcade-Cave_20210115.rbf. I tried it but it still occasionally freezes on boot. It seems to work well once it booted successfully.

**Comment by nullobject** [comment]
Thanks guys.

The main I wanted to know is what happens if/when the core freezes with 20210115: **does the green LED stay on, or go out**?

**Comment by johnklax** [comment]
Cave_20210115.rbf works for me too -- I had to change my vsync_adjust to 0 to get it to display on my TV but core/game load now. Will keep trying to see if I can get it to freeze and take a look at the LED....

**Comment by nullobject** [comment]
That's annoying, I didn't actually change anything to do with the behaviour :rofl:

**Comment by johnklax** [comment]
The core seems really solid for me. I've rebooted and reloaded a bunch of times, hasn't frozen yet.....

**Comment by bellwood420** [comment]
Disk LED result of Arcade-Cave_20210115.rbf:  When it freezes, Disk LED stays on.

I don't have IO board but I found that one of LEDs on DE10-Nano also showing Disk LED. 
Just in case I checked GPIO directly with multimeter, too.

**Comment by bellwood420** [comment]
I debugged Main using gdb. 

When it freezes, Main is trapped in infinite loop waiting for ACK from fpga.

https://github.com/MiSTer-devel/Main_MiSTer/blob/7c8d3cac6149ce2599355a212f67b6e5479df56d/fpga_io.cpp#L695-L704

Stack trace
```
0x00050a16 in fpga_spi (word=<optimized out>) at fpga_io.cpp:697
697                     gpi = fpga_gpi_read();
(gdb) bt
#0  0x00050a16 in fpga_spi (word=<optimized out>) at fpga_io.cpp:697
#1  0x0004ac5e in spi_w (word=<optimized out>) at spi.h:27
#2  spi_write (addr=addr@entry=0xb2471008 "\021", len=len@entry=4096, wide=<optimized out>) at spi.cpp:186
#3  0x0004cf54 in user_io_file_tx_data (addr=addr@entry=0xb2471008 "\021", len=len@entry=4096) at user_io.cpp:1393
#4  0x00079a16 in rom_finish (address=<optimized out>, send=<optimized out>) at support/arcade/mra_loader.cpp:344
#5  xml_send_rom (evt=<optimized out>, node=0xbeffe218, text=<optimized out>, n=<optimized out>, sd=0xbeffe274)
    at support/arcade/mra_loader.cpp:787
#6  0x00015d10 in _parse_data_SAX (in=in@entry=0x3bf338, in_type=in_type@entry=DATA_SOURCE_FILE,
    sax=sax@entry=0xbefffb34, sd=sd@entry=0xbeffe274) at sxmlc.c:1323
#7  0x0001606a in XMLDoc_parse_file_SAX (filename=filename@entry=0xbefffedf "/media/fat/_Arcade/DoDonPachi.mra",
    sax=sax@entry=0xbefffb34, user=user@entry=0xbeffe29c) at sxmlc.c:1622
#8  0x0007a1da in arcade_send_rom (xml=xml@entry=0xbefffedf "/media/fat/_Arcade/DoDonPachi.mra")
    at support/arcade/mra_loader.cpp:1045
#9  0x0004e68c in user_io_init (path=0xbefffea9 "/media/fat/_Arcade/cores/Arcade-Cave_DDR_20210115.rbf",
    xml=0xbefffedf "/media/fat/_Arcade/DoDonPachi.mra") at user_io.cpp:1030
#10 0x00071322 in main (argc=3, argv=0xbefffd94) at main.cpp:69

```

**Comment by jlrh** [comment]
> 
> 
> > Next experiment: this core will enable the green (disk) LED while the DDR3/SDRAM memory is waiting for data to be written. It should turn _off_ once the core has finished loading.
> > https://1drv.ms/u/s!AlvbZtqpJbLwgiWBWCI0rGuFRFGk?e=EgNgsD
> 
> This worked for me! I have the 128 Meg SD ram as well and was getting the freeze immediately with kbd mode stuck in upper left since yesterday after multiple tries. As soon as I added this rbf it worked great played through a few stages and everything seems to be working as it should. Thank you for working on this so quickly, really enjoying the core!

This worked for me now too!

32MB SDRAM, SDRAM and 57.5 options in advanced configuration and HDMI output. 

Thank you!!!!

**Comment by alienator303** [comment]
Using the latest DDR core , Arcade-Cave_20210115.rbf  , same freezes at start (but sometimes with garbage on screen) , all leds turned on.

the core however boots without the rom on my sd card (same for public release)

**Comment by nullobject** [comment]
@alienator303 Thanks, good to know about the LEDs.

**Comment by geekmiki** [comment]
As other have reported: Arcade-Cave_20210115.rbf doesn't solve the issue, still crashes with all leds on.

**Comment by nullobject** [comment]
Sure thanks, I haven't tried to solve it yet. I just wanted to identify where the issue may be with a bit of good old LED debugging :stuck_out_tongue_winking_eye:

**Comment by nullobject** [comment]
(It's difficult for me to find a bug that doesn't occur for me, so I'm just trying to get a few diagnostics from you guys).

**Comment by bellwood420** [comment]
Experimantal 1

I forced ioctl_wait = 0 and rebuild rbf.
Then it doesn't freeze but of course ROM doesn't seem to be loaded correctly.

Experimental 2

I connected ioctl_wait to LED_USER and rebuild rbf.
When it freezes, ioctl_wait stays high, which causes the infinite loop.
When it does not freeze, LED dims, which means ioctl_wait alternates between high and low so fast, then ioctl_wait turns low.

There seems to be something wrong with ioctl_wait signal supplied from the core.

**Comment by nullobject** [comment]
@iso70x That is defintely what it feels like. It could be a variance between SD card timing (or whatever), which is enough to make the bug appear for some folks and not others.

@bellwood420 Nice. Forcing `ioctl_wait` to zero is something I wanted to try, but waiting for the ROM to download is obviously required.

When `ioctl_wait` is asserted indefinitely, it makes sense that MiSTer freezes. I guess the question is: what is causing `ioctl_wait` to be always asserted? It must be a bug in my code somewhere :thinking:

**Comment by iso70x** [comment]
@nullobject If it is slight variance, it's so close in my case that just adding this LED debug feature has been enough to push it into the functional range on my system.  😄 I just tried the core a few more times and it launches perfectly.  
Oh, and LED goes off after core has loaded btw.

**Comment by alienator303** [comment]
> 
> 
> The diagnostic core released yesterday (Arcade-Cave_20210115.rbf) with the LED troubleshooting feature works 100% on my MiSTer (which I realise makes zero sense as no actual meaningful code was changed, only the LED debug feature). The public release fails 100% of the time, despite memory config, rom version, refresh/vsync (I've tried literally everything that's been suggested and the result remains the same).
> 
> I did the following to test this:
> Downloaded the new debug core from the onedrive link in earlier post.
> I currently have a copy of the 'no warn' mra called 'DoDon-test.mra' using this line:
> ` <rbf>Arcade-Cave</rbf>`
> (so it's using the new "Arcade-Cave_20210115.rbf"). When I launch using this mra, the game runs 100% successfully. I just tried it 5 times in a row and it's not failed a single time.
> 
> I then tried the original 'no warn' mra which uses the public core and has:
> ` <rbf>cave</rbf>`
> This locked up immediately.
> 
> I realise how frustrating this must be to debug given it can't be reproduced consistently, and I've also noticed that while this new version of the core is solving it for some people (myself included), it's probably not fixing it in all cases.
> 
> Given some people are reporting the issue is intermittent on their systems also, could this be a timing issue during rom load that is right on a limit somewhere?

tried this with public rbf, and last 2 test rbf, still locks up immediately :(
tried it again with beta 6 and it works! 

aargh! :D

**Comment by nullobject** [comment]
@bellwood420 If I were to send you a signaltap config, do you think you could run a trace of the frozen core for me?

**Comment by bellwood420** [comment]
@nullobject 

Of cource I can, but I don't know how to load rom from mra with SignalTap. 
I can flash sof and trace but have no idea about it. Would you tell me?

Btw, I found out something new.
It seems that the issue is related to not SDRAM but DDR3.

In current your code, download waitReq is OR of DDR3's and SDRAM's.

https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/blob/c4a884b2479b7c7dcb89b50f0928020e7f04837c/src/main/scala/cave/MemSys.scala#L168

I tested by not using OR but building separately.
The result is: 
- Using DDR3 and DDR3's waitReq causes freeze sometimes. 
And when it freezes, ddrDownloadCache.io.out.waitReq stays high (checked by LED).
So FSM in Cache may be trapped in evict or fill state.
- Using SDRAM and SDRAM's waitReq does not cause freeze as far as I have tested.

At least, OR of waitReq is causing the issue, even SDRAM is selected.

Hmm. What is happening ahead of ddrArbiter then?

**Comment by johnklax** [comment]
> @nullobject If it is slight variance, it's so close in my case that just adding this LED debug feature has been enough to push it into the functional range on my system. 😄 I just tried the core a few more times and it launches perfectly.
> Oh, and LED goes off after core has loaded btw.

My experience as well

**Comment by ratherDashing** [comment]
> > @nullobject If it is slight variance, it's so close in my case that just adding this LED debug feature has been enough to push it into the functional range on my system. smile I just tried the core a few more times and it launches perfectly.
> > Oh, and LED goes off after core has loaded btw.
> 
> My experience as well

As a fellow developer this infuriates me to say, but yes, it works now.  This is one of those times where you put a comment above the log line that says

// Removing this log line will cause issues

And two years later someone down the line removes the darn log line and you get a P0.

**Comment by nullobject** [comment]
@bellwood420 Excellent, let's get you set up.

> Of cource I can, but I don't know how to load rom from mra with SignalTap.
> I can flash sof and trace but have no idea about it. Would you tell me?

When you are running a core directly from Quartus/SignalTap, you obviously can't load a MRA file. In this case you must place a boot ROM file on your MiSTer located at `/media/fat/games/cave/boot.rom`. When the core boots, it will automatically download this ROM file. It is nothing too special, just an uncompressed file containing the ROMs concatenated in the correct order (the same as defined in the MRA).

I have shared the boot ROM here: https://1drv.ms/u/s!AlvbZtqpJbLwgifdmD-vnquNFSj2?e=99Q4p1

I have also [created a branch](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/tree/debug-ioctl-wait) where the latest commit adds a SignalTap file. If you could run this to capture a trace of what is happening when the core freezes it would be super helpful for me.

I especially want to know which part of the DDR arbiter is causing the DDR wait. Is it the DDR memory itself, or is it one of the other devices connected to it?

Let me know how you go.

**Comment by bellwood420** [comment]
@nullobject 

Thanks for precious debugging information!

Now, I've got values while freezing , please take a look.
Trigger does not occur somehow, so no information about transitions obtained.
There seems to be no change in values while freezing except `emu:emu|Main:main|io_ddr_addr[31..0]`.
Every time I run analysis while freezing, I get the following values.
`02400000h,  02480000h, 02500000h`

![data](https://user-images.githubusercontent.com/50853463/104807570-45053180-5823-11eb-9a84-8c218379c387.jpg)

**Comment by nullobject** [comment]
@bellwood420 You are a legend :clap: Thanks so much for capturing that.

**Comment by nullobject** [comment]
@bellwood420 I pushed up a change to the SignalTap file to add a few signals from the download cache module.

When you get a chance would you please be able to run another trace to capture those new signals. I want to drill into what is happening.

In particular, I want to know what happens just before the `emu:emu|Main:main|MemSys:mem|MemArbiter:ddrArbiter|io_in_0_rd` signal is asserted.

**Comment by nullobject** [comment]
@bellwood420 I pushed up a change to the SignalTap file to add a few signals from the download cache module.

>Every time I run analysis while freezing, I get the following values.
> 02400000h, 02480000h, 02500000h

That is the video DMA beginning a read from the frame buffer. Those addresses (e.g. `02400000h`) are the start addresses of the different pages in the frame buffer (e.g for double/triple buffering, etc.).

It appears that it is getting stuck during that first read from the frame buffer, which is blocking all other DDR memory requests (including writing the downloaded ROM data).

Could you please pull the latest commit from the branch and run another trace?

I want to see what happens during the first frame buffer read. Is the DDR memory even responding? Or is my DDR controller not counting the bursts properly and waiting indefinitely for a transfer that has already finished?

**Comment by bellwood420** [comment]
@nullobject Before pulling latest commit, I would like to upload capture which I just got.

vcd and jpg included
[Bad..zip](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/files/5824315/Bad.zip)

**Comment by nullobject** [comment]
@bellwood420 Thanks, that is so awesome. My latest theory is that it's possibly an initialization issue.

Perhaps if the first DDR request (for the frame buffer read) occurs too soon, it doesn't get processed properly?

Would love to see what is happening with that first DDR request :eyes:

**Comment by bellwood420** [comment]
@nullobject io_ddr_rd cannot be triggered as well as io_ddr_wr when it freezes.

**Comment by bellwood420** [comment]
@nullobject 

To see what is happening at the very beginning of boot, I changed trigger to Power-Up Trigger. 
As a result, after so many numbers of booting from both Signal Tap and rbf, it hasn't caused freeze yet for better or worse.
All the rbf before the change still causes freeze.

Note that changing to Power-Up Trigger requires re-compile, which means signal routing changes for that.
And now, after many attempts, I noticed that the probability of freezing seems to increase as the MiSTer warms up.
Btw my MiSTer doesn't have a fan. It's just bare DE10-Nano with a copper heat-sink.

IMHO, timing issue may be breaking essentially correct logics, but just an estimation.

**Comment by MittyTheRall** [comment]
updated MiSTer through the update_all script
DoDonPachi would hang forever on KBD Mode: Joystick 1
can confirm Cave_20210115.rbf fixed the issue and I can now load the game.

**Comment by bellwood420** [comment]
@nullobject  I did some experiment.

I covered MiSTer with cloth to overheat it. 
Result: 
- One without PowerUp Trigger causes freeze all the time from at some point.
- One with PowerUp Trigger doens't cause freeze.

Next, I put an ice on heatsink to cool it down.
Result: 
- One without PowerUp Trigger stops freezing gradually.
- One with PowerUp Trigger doens't cause freeze.

Yes, temperature matters and timing issue must exist somewhere on booting process.
With using the hypothesis, we can understand random phenomena in different devices and builds.

**Comment by nullobject** [comment]
@bellwood420 Wow, thanks for the thorough investigation.

I'm about to sit down and spend a few hours on the problem. I was looking through the initialisation code last night and I suspect there could be some kind of race condition there.

I think it is possible a DDR3 request could have been started, *before* the rest of the core has been reset properly. If a DDR request is already in progress when the frame buffer tries to perform a read, then it probably will never be completed and thus the whole download process will hang.

It's very interesting with what you observed with the different temperatures...maybe nudging the race condition one way or the other 🤔

**Comment by nullobject** [comment]
@bellwood420 I have made some changes to the handling of reset signals during initialization.

Will you please try it: https://1drv.ms/u/s!AlvbZtqpJbLwgijX3WzbTufAqD74?e=dgkTMW

**Comment by bellwood420** [comment]
@nullobject 

0117 still sometimes(but not always) causes freeze in warmed condition, where 0115 always causes freeze.
At least, something has changed.

**Comment by nullobject** [comment]
@bellwood420 Ok, thanks for testing. Will keep looking...

**Comment by bellwood420** [comment]
@nullobject 

I tried adding the following line to qsf as reset_sys_1 has about 290 fan-out to various modules.
`set_instance_assignment -name GLOBAL_SIGNAL "GLOBAL CLOCK" -to emu|reset_sys_1`

Experiment1
- Applying to 39b8bdede216152802566db83b0f44444cc1b8ea 
- Before: Always freezes
- After: Sometimes freezes

Experiment2
- Applying to 1a001bc9d0abbe9802b4b1d5a22a2d0418f436c3
- Before: Sometimes freezes
- After: Sometimes freezes (seems no change)

The result shows the line has effect on freezing but does not fix the issue completely. 
Also troubling, timing violation to Main|SDRAM module is emitted by Timing Analyzer due to the change.

**Comment by bellwood420** [comment]
@nullobject

Changing synchronous reset to asynchronous reset seems to lower the probability.

```
class Main extends Module with RequireAsyncReset {
```

But after tens of reloading, finally I encountered freeze...

**Comment by nullobject** [comment]
@bellwood420 I just saw it happen myself, but after setting up a signaltap to try to capture some signals it no longer happens!

OMG, this is very frustrating! :shrug:

**Comment by bellwood420** [comment]
@nullobject This is exactly caused by some timing issue in reset net, affected by routing, temperature and device dispersion.
Why does not Timing Analyzer in Quartus report it? It's troubling.

**Comment by nullobject** [comment]
@bellwood420 Hey, so I just captured a trace where the `io_ddr_rd` signal was asserted during boot, but the DDR never responded. This is interesting.

It seems like my arbiter and DDR controller are doing the correct things. I wonder if I need to wait longer before issuing the first DDR request during boot?

**Comment by bellwood420** [comment]
@nullobject That's nice and interesting. How about pending request for the frame buffer for a while?

**Comment by nullobject** [comment]
@bellwood420 Sure. Maybe I could prevent the video DMA from making any DDR requests until after the ROMs have been downloaded?

**Comment by calle81** [comment]
Just wanted to report that latest core "Arcade-Cave_20210117.rbf" is the first build that works on my mister. Huge thanks! 👍

**Comment by nullobject** [comment]
@calle81 Awesome. I don't think it's quite there, but we're getting close 😀

**Comment by bellwood420** [comment]
@nullobject I 'd love to test your new one :)

I tried to improve reset synchronizer by adding attribute and increasing number, but the result remained the same. 

Now I noticed that I was digging into a wrong direction.
Timing issue exists for sure as freezing depends on temperature.
However, it doesn't seem to be related to reset signal itself. I was misled by reset.
All my attempts had just changed fitter routing on chip...

**Comment by TheOpponent** [comment]
I can also confirm that 20210117 build works with my 128 MB SDRAM, while the current release 20210113 does not.

**Comment by mfncl9991** [comment]
20210117 not working for me, freezes at the same screen. With the chat about thermals possibly being an influence, my MiSTer is passively cooled through a misteraddons aluminum case so probably runs a tad hotter than most.

**Comment by bellwood420** [comment]
@nullobject  I'd like to share the result of my further attempt.

Conditions:
- Based on edc5bce55826c24a606595561d597a63819da146 
- Added D-FF on DDR3 bus between emu port and your Main core port 
- Changed compile setting to Aggressive

Results:
- Not freeze at booting and loading rom
- But freeze at WARNING message with high probability
- The point of freeze is random
- Rarely passing through WARNING message

Btw is 753bb17b31d5ad28892c7b93f79ee63791adb1a0 ready for me to test?
Thanks for intensive work.

**Comment by nullobject** [comment]
@bellwood420 I fixed a clock domain crossing issue in the `VideoFIFO` module.

I believe that this could have caused metastability in DDR read signal. Let me know if this improves things: https://1drv.ms/u/s!AlvbZtqpJbLwgiz8kDxJD1ROBZIE

**Comment by nullobject** [comment]
To make things worse, my MiSTer decided to give up the ghost and isn't working anymore :ghost:

**Comment by bellwood420** [comment]
@nullobject OMG. I experienced freezing 2 times in a hour unfortunately. But it is low probability.

**Comment by nullobject** [comment]
> OMG. I experienced freezing 2 times in a hour unfortunately. But it is low probability.

@bellwood420 Was that during boot, or while the game is running?

**Comment by bellwood420** [comment]
@nullobject It was on booting as well as before.

**Comment by nullobject** [comment]
Ok, here's one more (unfortunately I can't test): https://1drv.ms/u/s!AlvbZtqpJbLwgi1ddmAhsOvcX1d8?e=KUz5zl

I have updated the DDR controller to latch the request, rather than passing signals through.

P.S. Someone is sending me a new DE10-Nano express post, so I should be back up and running in a couple days.

**Comment by bellwood420** [comment]
> Ok, here's one more (unfortunately I can't test): https://1drv.ms/u/s!AlvbZtqpJbLwgi1ddmAhsOvcX1d8?e=KUz5zl
> 
> I have updated the DDR controller to latch the request, rather than passing signals through.

@nullobject It outputs corrupted video. Should I keep loading to see if it freezes?

![IMG_0132](https://user-images.githubusercontent.com/50853463/104876826-c7b0fc80-599b-11eb-87d1-6b3dffb78df8.jpeg)

**Comment by nullobject** [comment]
@bellwood420 I guess it will be interesting to see whether it freezes, although I obviously mucked something up.

Will have to wait until I get my new MiSTer I guess...to difficult to "fly blind".

**Comment by MJY71** [comment]
I get the same scrambled video as well with that most recent test build.

**Comment by bellwood420** [comment]
@nullobject It hasn't been causing freeze so far.
I don't know whether it is just a matter of probability, but it seems to be lower than before.

Of course it is so difficult to debug without your device.
I hope you will get a new one soon and hope mine will not pass away.

**Comment by alienator303** [comment]
Core Arcade-Cave_20210118.rbf is the first core that boots again here after beta 6... so , success !!! however, i also get the scrambled graphics/output as posted here above =)

**Comment by deportivista** [comment]
I have solved this issue unplugging the controllers. For me it is booting now only with keyboard plugged. Then you can plug the controller and configure it as usual in OSD into the game

**Comment by bellwood420** [comment]
> @bellwood420 Hey, so I just captured a trace where the `io_ddr_rd` signal was asserted during boot, but the DDR never responded. This is interesting.
> 
> It seems like my arbiter and DDR controller are doing the correct things. I wonder if I need to wait longer before issuing the first DDR request during boot?

@nullobject I captured an interesting trace (when it freezes) where the `io_ddr_rd` signal was asserted during boot, and the DDR **did respond** to burst read, but arbiter selected 0x08 only during `io_ddr_rd` was asserted, resulting DDR valid output to be ignored.

[bad2.zip](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/files/5833617/bad2.zip)

Note that this is captured on `debug-ioctl-wait` branch (at commit 98c8a52628439e95e65460d000b6ab58b1e52555).

Please just ignore it if this behavior would be fixed by measure to VideoFIFO CDC issue on `master` branch(127849d74d74003857865fc4c8d429e16dafd69f)

**Comment by nullobject** [comment]
@bellwood420 Thanks. That is a very interesting trace.

It's sightly concerning because the arbiter is making a request *without* any input requests. Maybe there's a bug in the arbiter? :eyes:

**Comment by bellwood420** [comment]
@nullobject I didn't notice that!

![compare](https://user-images.githubusercontent.com/50853463/104989837-ac5af580-5a5e-11eb-8f70-c737c717fcd8.png)

🤔

**Comment by bellwood420** [comment]
I've been trying various builds. It's getting confusing. To make things clear, I checked out the latest commit(2b521d113c5cc1aeb5575e6055694b87b865a255) of `master` branch and built it.
I experienced freeze. So the issue remains.
I wish I could get a trace with it...

**Comment by bellwood420** [comment]
@nullobject

**New experiment:** Forcibly changed clock_sys to 50 MHz

- Based on 2b521d113c5cc1aeb5575e6055694b87b865a255
- No boot freeze so far (1 hour endurance repeated loading)
- Rom seems to be loaded to the end according to LED and UART log
- But game does not start with black or noise screen (nothing is written to frame buffer?) 
Of course, 50 MHz should not be targeted by the design, so it's not surprising.

Btw, with original 96 MHz, I encounter boot freeze once in 15-30 min during repeated loading.

**Comment by IcarusFW** [comment]
From my initial testing of the Jan 13th core, the game takes a considerable amount of time to get past a black screen, before going straight to a screen with rainbow noise on it if it's the first time the core has been run. If I change the core to use DDR3 instead of SDRAM, then reset, the game still takes a bit of time to get past the black screen, but then loads into the boot sequence correctly.

Tested the Japanese .mra and it's playable after doing this. The "no-warn" .mra hangs permanently on the black screen and requires a power cycle to get the MiSTer back to a usable state. (I use a MiSTer with a 128MB RAM module.)

**Comment by nullobject** [comment]
@bellwood420 That is interesting with the clocks.

I have increased the video clock speed to 26.4MHz (4x) to fix another issue, so it will be interesting to see whether this has any affect:

https://1drv.ms/u/s!AlvbZtqpJbLwgjRZ16eGMm5GSf2c?e=NRUmDg

**Comment by bellwood420** [comment]
@nullobject Unfortunately it freezed on boot as before...  It was 30 minutes after I started repeated trying.

**Comment by alienator303** [comment]
> https://1drv.ms/u/s!AlvbZtqpJbLwgjRZ16eGMm5GSf2c?e=NRUmDg

this one now works perfectly on my mister.. yay!!

**Comment by bellwood420** [comment]
@nullobject I did another experiment. With clock_sys 80 MHz,  game can start, but the boot freeze exists.

Hmm...

**Comment by ae86gts** [comment]
> > https://1drv.ms/u/s!AlvbZtqpJbLwgjRZ16eGMm5GSf2c?e=NRUmDg
> 
> this one now works perfectly on my mister.. yay!!

Same here. Works great now.

**Comment by bellwood420** [comment]
@nullobject I have an question.

On commit be3ac6b76bee38f0ca0e78085366785e2d4e243d, with sof+boot.bin way, game does not start even though rom loaded successfully (just like my low clock experiment). 
No problem with rbf way (boot freeze not checked yet). 
No problem with former commit (for example, 2b521d113c5cc1aeb5575e6055694b87b865a255)

Is my boot.bin outdated or some other issue happenning?

**Comment by mfncl9991** [comment]
> @bellwood420 That is interesting with the clocks.
> 
> I have increased the video clock speed to 26.4MHz (4x) to fix another issue, so it will be interesting to see whether this has any affect:
> 
> https://1drv.ms/u/s!AlvbZtqpJbLwgjRZ16eGMm5GSf2c?e=NRUmDg

+1 - also working for me (at least for all of the four times I booted the core)

**Comment by bellwood420** [comment]
@nullobject I'd like to report the latest interesting result.

- Based on be3ac6b76bee38f0ca0e78085366785e2d4e243d
- Compared 2 loading methods
- sof+boot.rom : No boot freeze observed so far
- mra+rbf :  Boot freeze observed rarely

Something is different in boot sequence between the two.
Maybe duration between reset deassertion and rom load starting?🤔

Btw, this time, the result was obtained by automatically.
I was tired of repeating loading many many times, so I made a tool.
Now I can do other things with a cup of coffee :)

- sof+boot.rom auto loading tester
https://gist.github.com/bellwood420/12f56dc551c205f8c649f1b77466392b
Flash device with quartus_pgm and determine the result from UART output 

- mra+rbf auto loading tester
https://gist.github.com/bellwood420/41cb62f5e36c80b51f230bd6b452a1fb
Manipulate OSD menu via mbed uart2keyboard bridge(below) and determine the result from UART output 

- mbed uart2keyboard bridge
https://gist.github.com/bellwood420/d1ae2c02e2e5dac9d725d5e939ec9343

The log of the tool is as follows:
- sof+boot.rom
```
Test started at 2021/01/21 15:04:38
Test ended at 2021/01/21 17:58:57
Total attempts:  409
Boot freeze:     0
Flashing issue:  0
```
- mra+rbf
```
Test started at 2021/01/21 21:00:07
Boot freeze detected at 2021/01/21 21:14:45
1 of 92
Boot freeze detected at 2021/01/21 21:28:37
2 of 176
Boot freeze detected at 2021/01/21 21:54:19
3 of 336
Test ended at 2021/01/21 22:00:25
Total attempts:  372
Boot freeze:     3
```

**Comment by coderkind** [comment]
> @bellwood420 That is interesting with the clocks.
> 
> I have increased the video clock speed to 26.4MHz (4x) to fix another issue, so it will be interesting to see whether this has any affect:
> 
> https://1drv.ms/u/s!AlvbZtqpJbLwgjRZ16eGMm5GSf2c?e=NRUmDg

How are updates/releases coordinated in relation to MiSTer update scripts? This change won't come in as part of running the official MiSTer update script unless part of a new core release, or MiSTer update? I couldn't find any references to "cave" in [theypsilon's combination update script](https://github.com/theypsilon/Update_All_MiSTer) either (what I usually run), so the only other alternative to trying any of these updates is replacing individual files with ones posted here?

**Comment by owlnonymous** [comment]
Arcade-Cave_20200120.rbf and MRAs fixes my boot hangup. Thanks!

**Comment by IcarusFW** [comment]
Tried the Jan 20th beta and associated MRAs from the link above, still getting hard bootlock issues with:
- Dodonpachi
- Dodonpachi (no warn)
- Dodonpachi (Arrange)
The Dodonpachi (Japan) MRA only works as it was previously set to DDR3 from SDRAM during a previous test, and even then it is occasionally fussy about it and hangs, requiring a soft reset for it to work.

**Comment by MJY71** [comment]
Just thought I would add that the Arcade-Cave_20210120.rbf is booting and playing just fine for me as well so far, with almost all of the current DoDonPachi MRAs. However, I found that I do have to add the <skip CRC> patch to the main DoDonPachi MRA (International, Master Ver. 970205) to get that one to boot for some reason... the same patch found in the (No Warn) MRA.

**Comment by bellwood420** [comment]
@nullobject New result.

- Based on fa573fcc878f1d97aacb4595a7cf39f0c540e6c7
- mra+rbf

```
Test started at 2021/01/23 12:57:07
Boot freeze detected at 2021/01/23 13:08:18
1 of 35
Boot freeze detected at 2021/01/23 13:20:37
2 of 72
Boot freeze detected at 2021/01/23 13:48:18
3 of 158
Test ended at 2021/01/23 13:59:24
Total attempts:  192
Boot freeze:     3
```

**Comment by nullobject** [comment]
@bellwood420 Thanks for testing it again.

> I tweaked Main_MiSTer so that HPS reboots after loading rbf (and Main_MiSTer loads mra even after reboot).
> The original behaviour is that Main_MiSTer process restarts itself after loading rbf.

How did you do this?

**Comment by nullobject** [comment]
How about this build (593288e)? https://1drv.ms/u/s!AlvbZtqpJbLwgjVHUBkSQ2iB-6Ia?e=3BjH7E

I have disabled the frame buffer and video DMA until _after_ the ROM has finished downloading.

**Comment by bellwood420** [comment]
> @bellwood420 Thanks for testing it again.
> 
> > I tweaked Main_MiSTer so that HPS reboots after loading rbf (and Main_MiSTer loads mra even after reboot).
> > The original behaviour is that Main_MiSTer process restarts itself after loading rbf.
> 
> How did you do this?

@nullobject Just tweaked its source and rebuild.

[src.zip](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/files/5859973/src.zip)

Please get the diff from original.

**Comment by bellwood420** [comment]
> How about this build ([593288e](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/commit/593288e5c7fcde6c2f4ae862218a530d637e3729))? https://1drv.ms/u/s!AlvbZtqpJbLwgjVHUBkSQ2iB-6Ia?e=3BjH7E
> 
> I have disabled the frame buffer and video DMA until _after_ the ROM has finished downloading.

@nullobject I just started auto testing. Please wait until done. It takes a bit as you know.

**Comment by bellwood420** [comment]
@nullobject 

At first, it was going very well, and I thought about telling you a good news.
But... Arrrrgh :( 
Freezed... DISK and USER led was on. The same phenomena as before.

```
Test started at 2021/01/23 18:22:31
Boot freeze detected at 2021/01/23 19:52:22
1 of 237
Boot freeze detected at 2021/01/23 20:08:12
2 of 276
Test ended at 2021/01/23 20:08:21
Total attempts:  276
Boot freeze:     2
```

Btw, now I'm suspecting burst writing.

What will happen to f2sdram interface when the current core is forcibly reset by loading core from the OSD menu during burst writing.

**Comment by bellwood420** [comment]
@nullobject 

I have figured out a definite way to cause boot freeze.

No tricky way needed like tweaking Main_MiSTer or executing something like a command.
Just load Cave core from OSD menu in the core I've just made:

https://github.com/bellwood420/DDR3Burst_MiSTer

This core executes DDR3 burst write infinitely.

- When Cave core is loaded from this core, it freezes on boot.
- When some cores using FB on DDR3 for vertical video (like Arkanoid, Breakout...) is loaded from this core, it doesn't cause freeze but its video outputs black screen or garbage image remaining in memory.
Rotated horizontal video (not via FB) is fine.
- When cores not using FB is loaded from this core, there is no problem.

Forced resetting during burst writing must be the cause of freeze.
f2sdram interface between FPGA and HPS must be stucking due to this.

The reason of boot freeze happening only occasionally has to do with whether forced reset from HPS overlaps with DMA period or not.

Forced reset from HPS when core loading is inevitable.
So I think the only way to fix the issue by FPGA side is to change BURSTCNT = 1.
I hope it doesn't affect speed...

Or if there is any way to reset f2sdram interface from FPGA or HPS, that would be work.

**Comment by nullobject** [comment]
@bellwood420 Sorry for the delay, I've been preparing a conference talk I am giving today :sweat_smile: 

This is really great investigation, thank you.

Just to be clear, are you saying that if the DoDonPachi core is loaded while another DDR transaction is in progress, then it will freeze? Surely this is fairly common behaviour, I wonder why other MiSTer cores aren't suffering the same issue?

Unfortunately, it's not possible for me to use a burst size of 1 for everything (e.g. frame buffer DDR requests). There would be too much overhead in beginning 128 individual memory requests, rather than one big burst.

With a large burst size, you only have to pay the high latency cost **once**.

**Comment by bellwood420** [comment]
@nullobject  Good luck with your speech👍

> Just to be clear, are you saying that if the DoDonPachi core is loaded while another DDR transaction is in progress, then it will freeze? Surely this is fairly common behaviour, I wonder why other MiSTer cores aren't suffering the same issue?

It seems that forced interruption of burst writing might be the bad for f2sdram interface(breaking pipeline?)
I haven't checked precisely yet but interruption of burst READING doesn't seem to cause freeze to DoDonPachi core.

I haven't investigated other cores that use DDR3 uniquely (if you know, let me know please)
But at least the cores that use FB for vertical video use BURST_CNT=1 in their sys framework.

I think I need to look into ascal.vhd in framework sys which also uses burst writing.

> Unfortunately, it's not possible for me to use a burst size of 1 for everything (e.g. frame buffer DDR requests). There would be too much overhead in beginning 128 individual memory requests, rather than one big burst.
> 
> With a large burst size, you only have to pay the high latency cost **once**.

Of course, I get it. We need to find another way.

**Comment by bellwood420** [comment]
@nullobject 

Regarding ascal.vhd, I found that when reset is asseted, burst write transaction is interrupted and it continues outputting write = 1 after that.

To see the effect, I tweaked sys_top.sv to ensure that reset is asserted during burst writing, since it was hard to manually assert reset exactly during writing period which is intermittent and very short.

You can try out this version.
https://github.com/bellwood420/DDR3Burst_MiSTer/tree/break-scaler-ddr3

As a result, loading another core from this core, scaler video output corrupts like below!

https://user-images.githubusercontent.com/50853463/105662482-2aae1080-5f13-11eb-8783-ee89fd9afea8.mp4

Ah, I remembered this situation. Deja vu!
It was when I was reloading a core over and over again for developing it.
Rarely my HDMI output corrupted like this.
I always did reboot to recover not pursuing the cause.
Now I understand that when it happened, I loaded the core at the perfect(?) timing, as if I had hit the jackpot in slot machine!

Btw the port of f2sdram used for scaler's framebuffer(f2h_sdram0) is different from the one used for generic DDR3(f2h_sdram1) provided to emu module.
So broken f2h_sdram0 does not cause freeze in Cave core but video corrupts.

**Comment by bellwood420** [comment]
@nullobject

I opened PR, and it worked well in my environment.
No boot freeze detected after 400 times of loading🎉

```
Test started at 2021/01/26 13:27:53
Test ended at 2021/01/26 17:54:59
Total attempts:  400
Boot freeze:     0
```

**Comment by nullobject** [comment]
@bellwood420 This is fantastic :+1:

Let's move the discussion over to your PR...

**Comment by ItsAtomico** [comment]
> @bellwood420 That is interesting with the clocks.
> 
> I have increased the video clock speed to 26.4MHz (4x) to fix another issue, so it will be interesting to see whether this has any affect:
> 
> https://1drv.ms/u/s!AlvbZtqpJbLwgjRZ16eGMm5GSf2c?e=NRUmDg

This version fixed the boot lockup problem for me 👍

**Comment by same31** [comment]
I have the same issue, Kbd mode: Joystick 1 freeze on black screen, impossible to get the menu back and I have to shutdown
the MiSTER.

I use the direct video mode, RGB setup to connect to a PVM thanks to an HDMI to VGA adapter.
This is my config:
direct_video=1
composite_sync=1

Cheers!

**Comment by nullobject** [comment]
The booting issue should now be fixed thanks to much help from @bellwood420.

For anyone still following along, could you please verify the latest version still works for you: https://1drv.ms/u/s!AlvbZtqpJbLwgjcowRlTFUSgntxD?e=aBnEwq

**NOTE: You must run the update script on your MiSTer first.** 

This is to update the MiSTer menu core, which includes some changes that fix DDR memory issues while DoDonPachi (and other cores) is started.

**Comment by geekmiki** [comment]
Works fine here! (I was on the 20210120 version before that also worked well. Initial release crashed)

**Comment by same31** [comment]
I've tried the files you supplied in the one drive link and it works!
Thank you

**Comment by IcarusFW** [comment]
> For anyone still following along, could you please verify the latest version still works for you: https://1drv.ms/u/s!AlvbZtqpJbLwgjcowRlTFUSgntxD?e=aBnEwq

@nullobject A little update - this appears to have fixed most of the loading issues now for me, although it occasionally locks up if you switch between MRAs (for example, switching from International, to Arrange, to Japan versions as provided by your archive), and from cold boot. This still requires a reboot to fix, as the OSD is inaccessible. It's far less frequent than previous versions of the core, though (1-in-25 attempts now as opposed to 1-in-2 attempts like before).

Currently running the core in DDR3 mode and at 60Hz.

**Comment by DaveGeorge81** [comment]
Tried the latest unofficial build (20210128) and I finally got it working.

**Comment by alienator303** [comment]
yep working here

**Comment by coderkind** [comment]
This hasn't filtered through to the main MiSTer updated yet, right? Still won't boot for me.

**Comment by nullobject** [comment]
>This hasn't filtered through to the main MiSTer updated yet, right? Still won't boot for me.

@coderkind No, not yet. So the version I posted here is not working?

**Comment by PeteAnto** [comment]
Just a heads-up; I experienced the same boot issue until I took a closer look at my cores and realized that outside of the "arcade-cave" core, there was existed a "cave" core (no arcade prefix, which is why I didn't notice it at first), that must have been downloaded through the official update or something... Once I removed the "cave" core, keeping only the latest "arcade-cave" core, the game booted fine.

**Comment by coderkind** [comment]
> @coderkind No, not yet. So the version I posted here is not working?

@nullobject I tried an earlier bugfix attempt earlier this week but it didn't fix things for me. Would need to go through which files end up in what directories via the updater to properly test a proposed revision (which is why I was asking about the MiSTer updaters). I know others have mentioned stray "cave" directories ending up in conflicting locations, so would need to police that. Sorry; will dive into it in more depth hopefully this weekend.

**Comment by bellwood420** [comment]
@nullobject I have a question. I don't know this is related to the still reported issue but let me ask.

> That is the video DMA beginning a read from the frame buffer. Those addresses (e.g. `02400000h`) are the start addresses of the different pages in the frame buffer (e.g for double/triple buffering, etc.).

`0x02400000` seems to be inside Linux memory map. Is it really safe?
 
My understanding is that above `0x20000000` (512 MB) is reserved for FPGA.
I'm sorry if I'm wrong or misunderstanding something.

**Comment by nullobject** [comment]
@bellwood420 Fair question.

This is the address that the `arcade_video.v` module uses: https://github.com/MiSTer-devel/Template_MiSTer/blob/b81bf87a553ff625244dc59c8fd93f81e36af31a/sys/arcade_video.v#L197

**Comment by bellwood420** [comment]
> @bellwood420 Fair question.
> 
> This is the address that the `arcade_video.v` module uses: https://github.com/MiSTer-devel/Template_MiSTer/blob/b81bf87a553ff625244dc59c8fd93f81e36af31a/sys/arcade_video.v#L197

@nullobject I think it is `0x24000000` not `0x02400000`

**Comment by nullobject** [comment]
@bellwood420 Oops, typo. Good catch.

I will check this :smile:

**Comment by bellwood420** [comment]
@nullobject 

To see the effect of conflicting memory, I did some experiment.

Since it is up to Linux kernel which physical address is allocated for each page, it is difficult to make confliction surely.

Therefore I limited Linux memory to 128 MB (originally 511MB) to reduce free memory area so that probability of conflicting will increase.

As a result, launching Cave core made MiSTer very unstable after ROM loaded.
Linux memory is corrupted by inappropriate memory access from FPGA.

**Comment by nullobject** [comment]
@bellwood420 That totally makes sense. I have fixed this in master.

I wonder if it was responsible for some of the unreliable boot issues? :thinking:

**Comment by Akaifutari** [comment]
Dangun Feveron doesn't boot, and the new Dodonpachi, Dodonpachi (Japan) and Dodonpachi (Arrange) neither.

Could be the memory module again?

**Comment by coderkind** [comment]
Can someone outline the testing procedure for giving recent core revisions a try? I did a complete clearout of my MiSTer's ROMs and set everything back to vanilla, then ran [theypsilon](https://github.com/theypsilon/Update_All_MiSTer) update script. I could see I got a /media/fat/_Arcade/cores/Cave_20210113.rbf as a result of running that.

I tried taking [Arcade-Cave_20210113.rbf](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/blob/master/releases/Arcade-Cave_20210113.rbf) from the `master` branch and uploading it to the same directory, and I still experience the freezing with/without the "Cave_20210113.rbf" also living in the same directory.

I've looked in the `master` release .mra files and can't see anything pointing to individual "20210113" version of .rbf files, so I'm not sure how you point to a particular bugfix candidate (sorry; I still need to read up on some of MiSTer's directory structuring). There's a few branches in the repo, so maybe I should be trying files from one of those? Otherwise, considering I've run the official updater from scratch (and the accompanying scripts in theypsilon's combined script), I imagine I'm getting [the latest release](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/releases/tag/20210113) as part of the update.

**Comment by coderkind** [comment]
> @nullobject
> 
> Hi!
> Dangun Feveron doesn't boot, and the new Dodonpachi, Dodonpachi (Japan) and Dodonpachi (Arrange) neither.
> 
> Could be the memory module again?
> 
> Edit: I removed the memory module from the Mister and it works perfect, even Dodonpachi cores.

Does the MiSTer require total removal from the main power, and/or removal of the RAM modules to properly reset? Is there any sort of RAM cleanup routine that can be run before a core is started? Wondering if I'm doomed to always experience the freezing because my 128MB RAM module froze with the initial release and now nothing I attempt (outside of removing the RAM modules) can get my MiSTer to properly deal with revised releases (like it's aggressively caching the initial failure)?

**Comment by bellwood420** [comment]
@coderkind 

In order to test the latest fix candidate, 

- Download `Arcade-Cave_20210128.rbf` and mra files from the following link provided by @nullobject on earlier post.

> For anyone still following along, could you please verify the latest version still works for you: https://1drv.ms/u/s!AlvbZtqpJbLwgjcowRlTFUSgntxD?e=aBnEwq

- Place `Arcade-Cave_20210128.rbf` on `/media/fat/_Arcade/cores/`

- Move-away existing Cave rbf (in your case, it will be `Arcade-Cave_20210113.rbf` or `Cave_20210113.rbf`) from `/media/fat/_Arcade/cores/`

- Update mra files

Note that:

- Fix candidate version is not officially released yet on the github repository, so update script cannot get it and you cannot get it from github repository as well. You need to get it from the OneDrive link which @nullobject providing and manually place it.

- mra does not point to specific date string like `20210113`, just points the string `cave` (`<rbf>cave</rbf>`). 
When mra is loaded, MiSTer tries to find files which have `.rbf` extension and start with `Arcade-cave_`, `Arcade-cave.`, `cave_`, `cave.` with case-insentive.
Then the last found file will be loaded.
Since the order of result generated by `readdir` function (used to iterate files)  is not guaranteed, the way to be sure is to move-away old Cave rbf files as @nullobject saying on ealier post

> Note: you will have to move the other Cave RBF files out of the way.

**Comment by Michael1260** [comment]
Thank you so much !!! It works now ! So happy to have one my best game on my Mister :-)

**Comment by felleg** [comment]
I haven't been able to read all the comments in this issue, yet, but I'd just like to point out that with the newest files downloaded from the Dangun Feveron Beta 2 (and having cleaned up all the old cores from my directories), I am having these issues:

- DoDonPachi standard boots without warning screen
- DodonPachi (no warn) doesn't boot
- DoDonPachi Arrange and Japan don't boot

On the flip side, Dangun boots just fine. :)

**Comment by bellwood420** [comment]
@felleg Have you updated mra files for DoDonPachi?

It seems that, at some point, a value for selecting game was added in mra.
Old mra without the value cannot boot DoDonPachi unless debug setting changed.
If you don't find the following part in mra, it is old and incompatible.
```
  <!-- select game -->
  <rom index="1">
    <part>1</part>
  </rom>
```

**Comment by TruREALigion** [comment]
> 
> 
> The booting issue should now be fixed thanks to much help from @bellwood420.
> 
> For anyone still following along, could you please verify the latest version still works for you: https://1drv.ms/u/s!AlvbZtqpJbLwgjcowRlTFUSgntxD?e=aBnEwq
> 
> **NOTE: You must run the update script on your MiSTer first.**
> 
> This is to update the MiSTer menu core, which includes some changes that fix DDR memory issues while DoDonPachi (and other cores) is started.

I made an account just to reply that this new core works.  I didn't try the MRAs

i tried everything the day your public core was released:
-keyboard freeze (green light on)
-tried deleting and re-adding roms (with the suggested MD5)
-tried DDR3, 60Mhz, etc (had to delete the roms just to get to the menus to change
-tried unplugged inputs and going keyboard only

basically gave up until @bellwood420 fix.  thanks everyone

**Comment by nullobject** [comment]
@TruREALigion Awesome. Good to know.

I'm not convinced it's 100% perfect, but I think it has been much improved. Will be cutting a new official release soon.

**Comment by brunolazzaro** [comment]
I had the same issue, although I managed to make it work once. Downloading the linked core in this thread did fix the issue!

**Comment by Michael1260** [comment]
> I had the same issue, although I managed to make it work once. Downloading the linked core in this thread did fix the issue!

Exactly, it works now with Arcade-Cave_20210128.rbf

**Comment by coderkind** [comment]
Thanks @bellwood420; tried updated files and all working well now.

**Comment by L3Encounter** [comment]
I'm still struggling guys. I've downloaded the files from Onedrive and dropped the new rbf into the roms folder and updated the mra files. All previous rbfs with a 'Cave' identifier have been deleted.
All I get back is white noise and notification that U7.bin is missing. Any thoughts?
Many thanks.

**Comment by L3Encounter** [comment]
Up and running now!!! Happy man 😎👍

**Comment by werpu** [comment]
No luck here, I had it occasionally running, with most of the times hanging at a black screen now tried, the files from 28 of January now can see the boot screen but it hangs there, but that reliably.
Same here 128MB ram module!
The normal Dodonpachi and dodonpachi japan mra file gives me a u7.bin not found, could be an issue with the mame file.

**Comment by werpu** [comment]
Ok seems to run as well for me, I had problems but deleting my mame folder and updating it with all the latest mame roms seems to have fixed dodonpachi for me too in conjunction with the testing build from 28.1

**Comment by piso77** [comment]
FWIW, i just did a system update and i'm hitting the "black screen Kbd mode: joystick 1" freeze too.

Haven't read the >100 comments yet, but what should i do to debug the issue?

_Arcade/cores/Cave_20210113.rbf
128MB SDRAM
md5sum 58a82638a2aa7e4e40b32bf365bff17d  Games/mame/ddonpach.zip

**Comment by anthonycaccese** [comment]
I am in the same situation as @piso77 and look to have a similar set up.  

Trying to run the latest core from the updater using the latest .mra and getting "black screen Kbd mode: joystick 1" freeze

_Arcade/cores/Cave_20210113.rbf
128MB SDRAM
Games/mame/ddonpach.zip (.217 version) md5sum 58a82638a2aa7e4e40b32bf365bff17d

**Comment by coderkind** [comment]
Is it worth committing [the fixes outlined above](https://github.com/MiSTer-devel/Arcade-Cave_MiSTer/issues/18#issuecomment-770325514) as an official core release? With update_all scripts that update a range of MiSTer-related cores and features, the manual fixes are overwritten... meaning you have to re-apply the fixes each time you update your MiSTer. I don't think anyone on this thread has experienced worse results using it (from what I can tell, it's fixed the issue for the majority regarding freezing boot-up)?

I see a commit to the repo from 14 days ago, but I've just run my update script and tried Dodonpachi and it freezes again, so I assume that commit was not connected to the fix here.

**Comment by Pasha-From-Russia** [comment]
> 
> Serial log shows:
> 
> ```
> arcade_root /media/fat/_Arcade
> mame_root /media/fat/_Arcade
> Using interleave: input 8, output 16
>         file: /media/fat/_Arcade/mame/ddonpach.zip/u27.bin, start=0, len=0, map(2)=10
>      0: file: /media/fat/_Arcade/mame/ddonpach.zip/u26.bin, start=0, len=0, map(2)=1
> Disable interleave
> 100000: file: /media/fat/_Arcade/mame/ddonpach.zip/u50.bin, start=0, len=0
> 300000: file: /media/fat/_Arcade/mame/ddonpach.zip/u51.bin, start=0, len=0
> 500000: file: /media/fat/_Arcade/mame/ddonpach.zip/u52.bin, start=0, len=0
> 700000: file: /media/fat/_Arcade/mame/ddonpach.zip/u53.bin, start=0, len=0
> 900000: file: /media/fat/_Arcade/mame/ddonpach.zip/u60.bin, start=0, len=0
> B00000: file: /media/fat/_Arcade/mame/ddonpach.zip/u61.bin, start=0, len=0
> D00000: file: /media/fat/_Arcade/mame/ddonpach.zip/u62.bin, start=0, len=0
> F00000: file: /media/fat/_Arcade/mame/ddonpach.zip/u6.bin, start=0, len=0
> 1100000: file: /media/fat/_Arcade/mame/ddonpach.zip/u7.bin, start=0, len=0
> file_finish: 0x1300000 bytes sent to FPGA
> ```
> 
> When it freezes, the last line (file_finish) does not appear.
> The last line is missing in @ItsAtomico 's log, too.
> 
> Something wrong is happening there.


Had the same thing, BUT I solved this by removing patch offsets
```
    <!-- skip CRC -->
    <patch offset="0x005404">00 60 0C 00</patch>
    <!-- skip warning -->
    <patch offset="0x0054E0">00 60 50 00</patch>
```
from mra file

It loaded but crashed after a minute and I was not able to run it again...

**Comment by BretYancy** [comment]
so strange, im now running into this issue with the 4/6/21 release.
DDP boots to a black screen.

**working setup for DDP**
Arcade-Cave_20210113.rbf
128 RAM
HDMI output
57hz
low lag vsync mode
DoDonPachi (no warn).mra
ddonpach.zip (md5 - 9f887a16b6645327384a6623cbebb807)

**non-working setup for DDP**
Arcade-Cave_20210406.rbf
128 RAM
HDMI output
57hz
low lag vsync mode
DoDonPachi (no warn).mra
ddonpach.zip (md5 - 9f887a16b6645327384a6623cbebb807)
ddonpach.zip (md5 - 58a82638a2aa7e4e40b32bf365bff17d)

ive tried the advanced core settings for PCB to use dodonpachi, changed 57hz to 60hz, tried DDR3 instead of SDRAM, but the 4/6/21 release always boots to a black screen.
if i switch back to the 1/13/21 setup above, the game boots just fine

**Comment by hyp36rmax** [comment]
> I had no problem running the previous betas but Beta 6 hard locks my MiSTer immediately at core launch - the "Kbd Mode" message in the upper-left corner stays visible and the controls are completely unresponsive. I have the hit the first button on the MiSTer to restart the machine.
> 
> I'm using a 128 meg SDRAM module that tests stable at 140mhz (but no faster). I'm using VGA output. I tried different vsync_mode options in mister.ini but the outcome is the same.
> ![85FYnMa8](https://user-images.githubusercontent.com/49319761/104115444-c4977b80-52cc-11eb-9f1f-f4e7c74e04b6.jpeg)

I'm getting a similar issue with the static screen, followed by a blank screen on my Samsung 60" 4K TV, it's almost as if the source wasn't available or HDMI was unplugged. However when connected to a 1080P Monitor DoDonPachi loads with no issues.

**Comment by bibou42** [comment]
@hyp36rmax I have the same issue than you on my 4K TV. The issue is the TV, not the core.
If I set the core at 60Hz, it works.
If you can't see anything, push the menu button, then down 6 times then Accept 2 times.

**Comment by bibou42** [comment]
I also recommend to use the last official released rbf with the last official mra.
Don't mix official released files with beta files.
One last thing, I seperate the _Arcade folder and the _Beta folder to be safe.

**Comment by hyp36rmax** [comment]
> @hyp36rmax I have the same issue than you on my 4K TV. The issue is the TV, not the core.
> If I set the core at 60Hz, it works.
> If you can't see anything, push the menu button, then down 6 times then Accept 2 times to switch to 60Hz.

OMG this worked! Thank you! Now that we know this can be reproduced I hope we will get a solution to this. Curious what caused this.

**Comment by bibou42** [comment]
Your TV can't display 57Hz.
There is no issue with the core.
It will work on every other screen accepting 57Hz.

**Comment by nullobject** [comment]
The boot issue should be fixed in the latest release. For everyone who was having a problem previously, please confirm it is now working.

**Comment by hyp36rmax** [comment]
> Your TV can't display 57Hz.
> There is no issue with the core.
> It will work on every other screen accepting 57Hz.

Thanks I understand this now.
